### DISCUSSION TITLE
Better extension ergonomics?

### DISCUSSION THREAD

**Discussion by vunhatchuong** [ibhagwan/fzf-lua#1628] [General]
Disclaimer: I'm a Lua and Nvim noob, all my experience with them is from configuring my dotfiles.

This might be a skill issue, but I find creating an extension for `fzf-lua` a bit harder compared to `Telescope`. Perhaps there’s a way to implement something similar to Telescope’s `register_extension`.

I created [fzf-lua-jumplist.nvim](https://github.com/vunhatchuong/fzf-lua-jumplist.nvim) based on [telescope-jumps.nvim](https://github.com/vunhatchuong/telescope-jumps.nvim/tree/feat/max_results-line_distance).

In `Telescope`, adding a new picker can be achieved cleanly with the following code:

```lua
local conf = require('telescope.config').values

pickers.new(opts, {
    prompt_title = 'Jumpbuff',
    finder = finders.new_table {
    results = get_jumplist(config),
    entry_maker = function(entry)
	    return {
		    value = entry,
		    display = entry.display,
		    ordinal = entry.display,
		    lnum = tonumber(entry.lnum),
		    filename = filepath,
		    bufnr = bufnr,
	    }
    end,
    },
    previewer = conf.qflist_previewer(opts),
    sorter = conf.generic_sorter(opts),
}):find()


return require('telescope').register_extension {
	exports = { jumpbuff = show_jumps },
}
```

---

Problem:

The challenge I face with `fzf-lua` is that the `Previewer` and `fzf_exec` are separate, yet they depend on each other **_line by line_**.

This often results in two common solutions:

1. Embed all necessary information in the `fzf_exec` text
    - This approach allows the `previewer` to extract the location of the selected text using regex.
      - Cons: The text fed into `fzf_exec` becomes cluttered, as you need to include additional data for the `previewer`. You’re stuck working with **_regex_**. :(

2. Store items globally using a dictionary
    - The `previewer` fetches the required data from a global dict.
        - Cons: You lose the natural ordering of items.

Like the disclaimer said I'm a noob so maybe there's another way.

**Comment by ibhagwan** [comment]
The 2 ways you’re describing are the most common, unfortunately attaching metadata is never going to be as seamless as in Telescope as you’re dealing with an external process, once the data is handed over to fzf all it knows is the string entry, there is no awareness of the lua context, so you’re left with 2 options:
1. Store metadata in the line and hide it with `--with-nth`
2. Store an index in the line (possibly hidden) with an additional structure (map) mapping the index to the metadata

**Reply by vunhatchuong** [reply]
Hmm... I'm spitballing here but Is it possible to make a wrapper `register_extension` function that takes in an `entry` list like `Telescope`?

```lua
function M.register_extension(input_list, fzf_opts)
    local text_list
    local preview_location_list

    for _, v in ipairs(input_list) do
        table.insert(text_list, v.text)

        table.insert(preview_location_list.{
            lnum = tonumber(v.lnum),
            col = v.col,
            filename = v.filepath,
            bufnr = v.bufnr,
        })
    end

    fzf_opts.previewer = preview_location_list
    require("fzf-lua.core").fzf_exec(text_list, {})
end
```

Since there must be a way `fzf-lua` tracks when you change an item, so maybe just +1 in the `previewer_list`?

**Reply by ibhagwan** [reply]
> Hmm... I'm spitballing here but Is it possible to make a wrapper register_extension function that takes in an entry list like Telescope?

Anything is possibles just gotta decide which previewer you want, the builtin or the native, better yet both, it would need some design and a few tricks perhaps but definitely possible.
